Added mesh triangulation methods, and the Triangulate Mesh node.

3DSinghVFX opened the pull request:
I have added the _Fan Span_ and _Ear Clipping_ methods for mesh triangulation. And added the _Triangulate Mesh_ node that allows triangulating the mesh,
In the advanced settings of the node, we can toggle between the _Fan Span_ and _Ear Clipping_ methods.
Example:

![Screenshot from 2020-04-18 23-03-17](https://user-images.githubusercontent.com/30294746/79644921-a97a5e00-81c9-11ea-9270-7bb88169ec00.png)

![Screenshot from 2020-04-18 23-06-14](https://user-images.githubusercontent.com/30294746/79644922-b0a16c00-81c9-11ea-80fd-99073fa05155.png)

I have tested the _Ear Clipping_ method on these meshes:
![79075117-5fe3cc00-7d0e-11ea-91ff-988592909fa8](https://user-images.githubusercontent.com/30294746/79645016-2c031d80-81ca-11ea-8b1a-35452a16d578.png)

OmarEmaraDev commented:
Can you provided the resources used to write this implementation? I am not familiar with the method myself, so if you have some good resources, I would appreciate it.

3DSinghVFX commented:
> Can you provided the resources used to write this implementation? I am not familiar with the method myself, so if you have some good resources, I would appreciate it.

References for _Ear Clipping_ method: https://arxiv.org/abs/1212.6038, https://www.coursera.org/lecture/computational-geometry/4-5-an-ear-cutting-algorithm-c3d7Y

These are important steps that I found during implementation:
1. Polygon should be in counter-close wise.
2. Then calculate the _inner_ angle at all vertices.
3. If a vertex says E is convex (small inner angle), and any other vertices do not lie in the triangle (E-1, E, E+1) formed by the vertex E. Then E is an _Ear_.
4. Then, store the triangle indices and remove the vertex E.
5. Repeat steps 2-4 until we left with only three vertices. 

Polygon Transform: I used _Rodrigues' rotation formula_ to rotate the polygon-vertices to XY-plane.

3DSinghVFX commented:
I have noticed that if we have a _weird topology_ mesh and triangulate it (using _Triangulate Mesh_ (except _Fan Span_ method which works in some cases) or _Triangulate Bmesh_ nodes (except _Fixed_ method which works in some cases)). It is possible that two triangle-polygons (or more) from different original polygons (of input mesh) can share the same vertices. Here are examples,

**Triangulate Mesh Node**:
![Screenshot from 2020-04-20 15-36-29](https://user-images.githubusercontent.com/30294746/79740273-08042100-831d-11ea-9ba5-ae296aa42110.png)

**Triangulate BMesh Node**:
![Screenshot from 2020-04-20 15-34-20](https://user-images.githubusercontent.com/30294746/79740290-105c5c00-831d-11ea-83b0-b10496da618a.png)

_To resolve this we have to either skip mesh validation, and the Blender does not care about the duplicate polygons (and can be removed with the Weld modifier) or add a method which removes the duplicated triangle-polygons._

**SkipValidation**
![Screenshot from 2020-04-20 15-44-28](https://user-images.githubusercontent.com/30294746/79740805-e2c3e280-831d-11ea-98e2-78811171a0eb.png)

JacquesLucke commented:
Unfortunately, we cannot allow duplicate edges/faces in a mesh. Blender's API docs are quite clear about that: https://docs.blender.org/api/blender2.8/bmesh.html#keeping-a-correct-state
Also `mesh.validate(...)` would mark such meshes as invalid.

Can you specify "a weird topology mesh" more precisely? Maybe we can check for this case efficiently before running the algorithm and then either discard such inputs or fix them afterwards.

OmarEmaraDev commented:
@JacquesLucke  I see. It is weird that the BMesh triangulate method returns duplicates in this case.

The topology Kuldeep is talking about is when two polygons share two edges, some triangulation algorithms will connect those two edges for each of the two polygons resulting in two duplicate triangles as illustrated here. (Left: Original polygons. Center: After triangulating one polygon. Right: After triangulating both polygons.)

![](https://user-images.githubusercontent.com/25300994/79858199-583bbb80-83cf-11ea-929f-02ab3687ee72.png)

JacquesLucke commented:
True, that seems to be a bug in Blender...
I think for now we could just do the validity check after running the algorithm. If the computed mesh is invalid, an empty mesh is returned instead and an error message is shown.

3DSinghVFX commented:
> True, that seems to be a bug in Blender...
> I think for now we could just do the validity check after running the algorithm. If the computed mesh is invalid, an empty mesh is returned instead and an error message is shown.

Thanks, @JacquesLucke and @OmarEmaraDev. I'll make the changes.

3DSinghVFX commented:
> Unfortunately, we cannot allow duplicate edges/faces in a mesh. Blender's API docs are quite clear about that: https://docs.blender.org/api/blender2.8/bmesh.html#keeping-a-correct-state
> Also `mesh.validate(...)` would mark such meshes as invalid.
> 
> Maybe we can check for this case efficiently before running the algorithm and then either discard such inputs or fix them afterwards.

@JacquesLucke and @OmarEmaraDev From the topology point of view, I think this situation may only occur when polygon vertices are not coplanar.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+    for i in range(polygonAmount):
+        polyLength = _oldPolyLengths[i]
+        polyStart = _oldPolyStarts[i]
+        neighbors = LongList(length = polyLength)
+        indices = LongList(length = polyLength)
```
Why do we need to store `neighbors` in an array? Aren't the neighbors already known from the original polygon indices?

3DSinghVFX replied:
Actually, if the input polygon has flip (has inverted normals) then we need to reverse the polygon indices. That is why I store in a `neighbors LongList`.

OmarEmaraDev replied:
Why not reverse the indices themselves? Also, what would happen if we didn't reverse?

3DSinghVFX replied:
The _inner angle_ and _convex_ methods need a counter-clockwise polygon. If the input polygon is clockwise then we have to reverse it to make it counter-clockwise.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+            neighbors.data[j] = _oldIndices[polyStart + j]
+            indices.data[j] = j
+
+        polyVertices = transformVerticesOfPolygon(oldVertices, neighbors)
```
Should we call this something more meaningful? Maybe `projectPolygonVertices`.

3DSinghVFX replied:
Done.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+
+                break
+
+            # Calculate inner-angle for all vertices of polygon
+            angles = LongList(length = amount)
+            for k in range(amount):
+                preVertex = polyVertices.data[indices.data[(amount + k - 1) % amount]]
+                earVertex = polyVertices.data[indices.data[k]]
+                nexVertex = polyVertices.data[indices.data[(k + 1) % amount]]
+
+                subVec3(&ab, &earVertex, &preVertex)
+                subVec3(&bc, &earVertex, &nexVertex)
+
+                normalizeVec3_InPlace(&ab)
+                normalizeVec3_InPlace(&bc)
+
+                angle = angleNormalizedVec3(&ab, &bc)
+                angles.data[k] = int((angle - angle % 0.001) * 180.0 / pi)
+            sortAngles = LongList.fromNumpyArray(numpy.sort(angles.asMemoryView()).astype(int))
```
Do we really need to recalculate everything everytime? I assume only two angles will change here. Also, using numpy here would be questionable. If you need to, use `qsort` from stdlib.

3DSinghVFX replied:
> Do we really need to recalculate everything everytime? I assume only two angles will change here.

That can be avoided.

> Also, using numpy here would be questionable. If you need to, use `qsort` from stdlib.
 
Okay. I did not know that.

3DSinghVFX replied:
Omar, I tried this but failed. Please help me,
```
from libc.stdlib cimport qsort

newAngles = LongList()
qsort(newAngles, newAngles.length, sizeof(int), compare)

cdef compare(const void * a, const void * b):
   return <int*>a - <int*>b
```

OmarEmaraDev replied:
`<int*>a` will cast `a` to an integer pointer, so you still need to dereference it to get the actual value `(<int*>a)[0]`.

3DSinghVFX replied:
> Do we really need to recalculate everything everytime? I assume only two angles will change here. 

I have modified the algorithm, now no need to recalculate everything everytime.

3DSinghVFX replied:
> `<int*>a` will cast `a` to an integer pointer, so you still need to dereference it to get the actual value `(<int*>a)[0]`.

I have done a test, but it is not working.
```
from libc.stdlib cimport qsort

qsort(newAngles, newAngles.length, sizeof(int), compare)

cdef compare(const void * a, const void * b):
   return (<int*>a)[0] - (<int*>b)[0]
```
![Screenshot from 2020-04-23 00-41-33](https://user-images.githubusercontent.com/30294746/80023491-3a6b7500-84fb-11ea-8769-c51e51713262.png)

OmarEmaraDev replied:
Looks like you need a pointer to the function. So try `qsort(newAngles, newAngles.length, sizeof(int), &compare)`.

3DSinghVFX replied:
Sorry for late reply. It gave the same error, `Cannot assign type 'object (*)(const void *, const void *)' to 'int (*)(const void *, const void *) nogil'`

OmarEmaraDev replied:
We forgot the function type:
```
cdef int compare(const void * a, const void * b):
```

3DSinghVFX replied:
Still the error is there,
![Screenshot from 2020-04-23 02-23-37](https://user-images.githubusercontent.com/30294746/80032924-832a2a80-8509-11ea-94ef-8ff9112f4751.png)

3DSinghVFX replied:
Is this related, https://stackoverflow.com/a/50777451

3DSinghVFX replied:
This code is working, 
```
from cpython cimport array
import array

cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    qsort(&a[0], amount, sizeof(int), &compare)
    print(b)

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```

OmarEmaraDev replied:
You need to release the GIL for the compare function. See https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil

3DSinghVFX replied:
Okay, but is it correct way to do that?

OmarEmaraDev replied:
It is in the signature of the expected callback, so I guess so.

3DSinghVFX replied:
Okay, 
```
cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    with nogil:
        qsort(&a[0], amount, sizeof(int), &compare)
    print(b)

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```

3DSinghVFX replied:
This is a complete function,
```
cdef quickSort(LongList newAngles):
    cdef Py_ssize_t amount = newAngles.length
    cdef array.array b = array.array('i', newAngles.asNumpyArray())
    cdef int[:] a = b
    with nogil:
        qsort(&a[0], amount, sizeof(int), &compare)

    cdef LongList sortAngles = LongList(length = amount)
    cdef Py_ssize_t i
    for i in range(amount):
        sortAngles.data[i] = b[i]
    return sortAngles

cdef int compare(const void * a, const void * b) with gil:
   return (<int*>a)[0] - (<int*>b)[0]
```
Speed testing:
**With qsort:**
![test1_gil](https://user-images.githubusercontent.com/30294746/80095907-5831ec00-8586-11ea-8a85-5336f50b8cee.png)

**With Numpy:**
![test1_nogil](https://user-images.githubusercontent.com/30294746/80095942-6bdd5280-8586-11ea-8905-2cdcbafe0295.png)

3DSinghVFX replied:
Can we use `sort` instead of `qsort` which is fast according this site: https://www.geeksforgeeks.org/c-qsort-vs-c-sort/?ref=rp

OmarEmaraDev replied:
I meant like this.

```cython
from libc.stdlib cimport qsort

cdef int compare(const void * a, const void * b) nogil:
   return (<int*>a)[0] - (<int*>b)[0]

qsort(newAngles, newAngles.length, sizeof(int), &compare)

```

3DSinghVFX replied:
It gave an error, maybe need to `newAngles` to C-object,
![Screenshot from 2020-04-23 17-40-47](https://user-images.githubusercontent.com/30294746/80097905-aa284100-8589-11ea-8bed-e3effc057054.png)

OmarEmaraDev replied:
`newAngles` is a `LongList`. Use `newAngles.data`.

3DSinghVFX replied:
Thank you so much for the help.

It is working But the output is not the same,
![Screenshot from 2020-04-23 17-50-38](https://user-images.githubusercontent.com/30294746/80098775-0f306680-858b-11ea-8bbf-567d0daa1487.png)

OmarEmaraDev replied:
A LongList is composed of longs, so I think `sizeof(int)` should be `sizeof(long)`.

3DSinghVFX replied:
Ahhh, it worked. 

Thank you so much :)

OmarEmaraDev replied:
How is the performance now? Is it better than numpy? I am not sure if we are allowed to use c++ here, so not sure about the STL sort function.

3DSinghVFX replied:
Yes, it is better.

But we use c++ in the Fast Noise.

3DSinghVFX replied:
Based on your suggestions, now performance is far better. For example triangulation of a sphere, timing reduces to `~17ms` to `1.45ms`.

3DSinghVFX replied:
Thanks. Okay.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+                v2 = polyVertices.data[earIndex]
+                v3 = polyVertices.data[nexIndex]
+                if convexVertex(v1, v2, v3) and not pointsInTriangle(polyVertices, v1, v2, v3, preIndex, earIndex, nexIndex):
+                    indices = LongList.fromNumpyArray(numpy.delete(indices.asMemoryView(), index).astype(int))
```
Same as I mentioned above.

3DSinghVFX replied:
Is there a way to delete an element Or should I use loop for that?

OmarEmaraDev replied:
The Animation Nodes list structures implement element deletion, so you can just do that. Either using the `del` keyword or the `removeValueAtIndex` method.
But if you can avoid deletions, that would be best. Perhaps by keeping track of what was removed without actually removing it.

3DSinghVFX replied:
I have modified the algorithm, now no need for deletion.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+    cdef Vector3 v1, v2
+    cdef Py_ssize_t i
+
+    for i in range(amount):
+        v1 = vertices.data[i]
+        v2 = vertices.data[(i + 1) % amount]
+        area += (v2.x - v1.x) * (v2.y + v1.y)
```
Instead of using modulo here. We can pull the last case out like this:

```python
    for i in range(amount - 1):
        v1 = vertices.data[i]
        v2 = vertices.data[i + 1]
        area += (v2.x - v1.x) * (v2.y + v1.y)
    v1 = vertices.data[amount - 1]
    v2 = vertices.data[0]
    area += (v2.x - v1.x) * (v2.y + v1.y)
```

3DSinghVFX replied:
Done.

OmarEmaraDev commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+        rotPolyVertices.data[i].y = rotPolyVertex.y
+        rotPolyVertices.data[i].z = 0
+
+    # Compute avg. vector
+    cdef Vector3 avgVector = toVector3((0, 0, 0))
+    for i in range(polyLength):
+        avgVector.x += rotPolyVertices.data[i].x
+        avgVector.y += rotPolyVertices.data[i].y
+        avgVector.z += rotPolyVertices.data[i].z
+    scaleVec3_Inplace(&avgVector, 1.0 / polyLength)
+
+    # Bring vertices of polygon to the world center
+    cdef Vector3DList offsetPolyVertices = Vector3DList(length = polyLength)
+    for i in range(polyLength):
+        offsetPolyVertices.data[i].x =  rotPolyVertices.data[i].x - avgVector.x
+        offsetPolyVertices.data[i].y =  rotPolyVertices.data[i].y - avgVector.y
+        offsetPolyVertices.data[i].z =  rotPolyVertices.data[i].z - avgVector.z
+
```
Is this really necessary? Can't we just skip it?

3DSinghVFX replied:
I have done testing, there is no longer need of this offset. I'll remove it.

3DSinghVFX replied:
Done.

OmarEmaraDev commented on animation_nodes/nodes/mesh/triangulate_mesh.py:
```diff
@@ -0,0 +1,35 @@
+        layout.prop(self, "methodType", text = "Use Advanced Method")
+
+    def execute(self, mesh):
+        meshOut = mesh.copy()
```
Use `dataIsModified` for the input mesh.

3DSinghVFX replied:
That gives this error,
![Screenshot from 2020-04-22 17-28-49](https://user-images.githubusercontent.com/30294746/79979232-d2973900-84be-11ea-8800-911ca02336d5.png)

OmarEmaraDev replied:
Can you show me the relevant code using `dataIsModified`?

3DSinghVFX replied:
This is edited _Traingulate Mesh_ node code,
```
import bpy
from bpy.props import *
from ... events import propertyChanged
from ... base_types import AnimationNode
from ... data_structures.meshes.validate import checkMeshData

class TriangulateMeshNode(bpy.types.Node, AnimationNode):
    bl_idname = "an_TriangulateMeshNode"
    bl_label = "Triangulate Mesh"
    errorHandlingType = "EXCEPTION"

    methodType: BoolProperty(name = "Use Advanced Method", default = False,
                             description = "Use advanced method for Mesh Triangulation",
                             update = propertyChanged)

    def create(self):
        self.newInput("Mesh", "Mesh", "mesh", dataIsModified = True)

        self.newOutput("Mesh", "Mesh", "meshOut")

    def draw(self, layout):
        layout.prop(self, "methodType", text = "Use Advanced Method")

    def execute(self, mesh):
        if self.methodType:
            mesh.triangulateMesh(method = "EAR")
        else:
            mesh.triangulateMesh(method = "FAN")

        try:
            checkMeshData(mesh.vertices, mesh.edges, mesh.polygons)
            return mesh
        except Exception as e:
            self.raiseErrorMessage(str(e))
```

3DSinghVFX replied:
This is the function in the `mesh_data.pyx` file
https://github.com/JacquesLucke/animation_nodes/blob/e2195df285ff9182e3b1cadbf7624ed9bfa8da9a/animation_nodes/data_structures/meshes/mesh_data.pyx#L181-L188

OmarEmaraDev replied:
I am not sure. I will have to investigate this myself. So just ignore it for now.

3DSinghVFX replied:
Okay.

3DSinghVFX commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
     return neighboursAmounts, neighboursStarts, neighbours, neighbourEdges
+
+def triangulatePolygonsUsingFanSpanMethod(PolygonIndicesList polygons):
+    cdef unsigned int *_oldPolyLengths = polygons.polyLengths.data
```
Okay.

3DSinghVFX replied:
Removed.

3DSinghVFX commented:
> Consider moving all of the functions into a separate file, maybe in `algroithms/mesh`.

In `algroithms/mesh_generation` or create a new folder `algroithms/mesh`?

3DSinghVFX commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+    cdef Py_ssize_t i, triAmount, polyLength
+
+    triAmount = 0
+    for i in range(polygonAmount):
+        polyLength = _oldPolyLengths[i]
+        if polyLength > 3:
+            triAmount += polyLength - 2
+        else:
+            triAmount += 1
```
Ahhh, I did not think about that.

3DSinghVFX replied:
Done.

OmarEmaraDev commented:
> In `algroithms/mesh_generation` or create a new folder `algroithms/mesh`?

Create a new directory.

3DSinghVFX commented:
> > In `algroithms/mesh_generation` or create a new folder `algroithms/mesh`?
> 
> Create a new directory.

Okay.

3DSinghVFX commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
+    return newPolygons
+
+@cython.cdivision(True)
+def triangulatePolygonsUsingEarClipMethod(Vector3DList oldVertices, PolygonIndicesList polygons):
```
Done.

3DSinghVFX commented on animation_nodes/data_structures/meshes/mesh_data.pyx:
```diff
@@ -370,3 +384,272 @@ def calculateLinkedVertices(int verticesAmount, EdgeIndicesList edges):
 
     return neighboursAmounts, neighboursStarts, neighbours, neighbourEdges
+
+def triangulatePolygonsUsingFanSpanMethod(PolygonIndicesList polygons):
```
The functions are moved to a separate file.

3DSinghVFX commented:
Hi @OmarEmaraDev 

> I just skimmed over the paper you used to implement the triangulation. Its seems their improved algorithm aims to reduced thin triangles by choosing the ear with the minimum angle and swapping edges when needed. It seems you implemented the former but not the latter, or did you?
 
I used minimum angle approach.

> Implementation wise, here are some points:
> 
>     * You copy and resort the whole angles array for every iteration which is bad, I think I mentioned this before, and while you only compute the two changed angles, we also need to avoid resorting the whole array. To solve this an incremental data structure could be used, like a Heap.

Okay. It computes only two angles i.e., angles of vertices adjacent to the ear-vertex. I do not know Heap method, but I'll try to implement it. 

>     * Shouldn't we skip reflex vertices in the `earNextNeighborIndices` function?

As mentioned in that paper, the removal of an Ear-Vertex can change the status of neighbour vertices (reflex or convex). To compute the angles of neighbour vertices (of Ear-Vertex), we need next-neighbour vertices whether they are reflex or convex.

>     * I don't see why conversion to degrees is necessary here, won't radians do just fine?

I added `(angle - angle % 0.001) * 180.0 / pi` to avoid the precision problem. If the double list with `(angle - angle % 0.001)` can handle that problem, we can remove degree conversion. 

>     * Can we compute directed angles and use them to identify reflex and convex vertices?

Sorry, what do you mean by _directed angles_? I had tried with polygon normal that did not work.



> 
> So generally, the structure should be as follows:
> 
>     1. Compute the **directed** interior angles and the ear status of the vertices. The sign of the angle indicates the reflex and convex status of the vertices.
> 
>     2. Construct a heap from the vertices.
> 
>     3. Extract the vertex with the minimum angle, do the triangulation.

During extraction, we have to check that the triangle formed by the Ear-Vertex with minimum angle does not include any other vertex. If it does then we have to go for next Ear-Vertex with minimum angle. 
 
>     4. Update the heap based on the two changed angles.
> 
>     5. Go back step 3 until done.

The overall structure of the current algorithm is similar to the above. We have to add the `Vertex Status` and `Heap` method.

Current algorithm:
0. Check a polygon (after projection) is counter-clockwise.
1. Compute the interior angles of each vertex.
2. Sort the angles.
3. Remove the Ear (convex, min interior angle) by the condition that formed triangle does not include any other vertex.
4. Update angles of neighbours of the Ear-Vertex.
5. Repeat step 2.

3DSinghVFX commented:
> > I used minimum angle approach.
> 
> So you are not doing edge swapping?

Yes. I'm not doing edge swapping.
> 
> > To compute the angles of neighbour vertices (of Ear-Vertex), we need next-neighbour vertices whether they are reflex or convex.
> 
> I am aware, but you are looping over all vertices, while you should only loop over convex vertices. Right?

Yes. If we store the Ear Vertex-Status (as 1) then we can just `continue` the loop for `reflex` vertices:
```
for i in range(amount):
    if vertexStatus[i] == 0: continue
    check for possible trianguation...
    update the angles and vertex-status
```
> > What do you mean by directed angles?
> 
> A directed angle is a signed angle. See the following diagram:
> 
> https://animation-nodes-manual.readthedocs.io/en/latest/user_guide/nodes/vector/vector_angle.html#rotation-difference

Ahh, that will work, we can compute angle and vertex-status simultaneously. Thanks for the info.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+        # Check polygon polarity.
+        if not polyCounterClockwise(verticesData):
+            for j in range(polyLength):
+                previousIndex = verticesData[j].previous
+                nextIndex = verticesData[j].next
+                verticesData[j].previous = nextIndex
+                verticesData[j].next = previousIndex
```
Put those in a function, just to make it clear what is happening.

3DSinghVFX replied:
It is added into `makeSurePolygonIsCounterClockwise(verticesData)` function.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+        # Calculate triangle polygon indices.
+        for j in range(polyLength - 2):
+            if polyLength - 2 - j == 1:
+                newPolyStarts[triangleIndex] = polyIndex
+                newPolyLengths[triangleIndex] = 3
+                triangleIndex += 1
+
+                index = 0
+                for k in range(polyLength):
+                    if verticesData[k].isEar:
+                        newIndices[polyIndex + index] = verticesData[k].index
+                        index += 1
+                        if index == 3: break
+                polyIndex += 3
+                break
```
It appears you are checking if this is the last triangle and skipping the extra checks if it is. What I would do is move this check after the loop. So you loop one time less `for j in range(polyLength - 3):` and then add the last triangle manually.

3DSinghVFX replied:
It is done.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+    previousIndex = verticesData[earIndex].previous
+    nextIndex = verticesData[earIndex].next
+
+    verticesData[earIndex].isEar = False
```
The ear is being removed, so you don't need to update any of its values. More on this later.

3DSinghVFX replied:
`verticesData[earIndex].isEar = False` is removed.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+    else:
+        verticesData[previousIndex].isEar = False
+
+    angle = calculateAngleWithSign(verticesData, nextIndex)
+    verticesData[nextIndex].angle = angle
+    if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, nextIndex):
+        verticesData[nextIndex].isEar = True
+    else:
+        verticesData[nextIndex].isEar = False
```
It seems we can move this to a function since it is repeated 3 times, 2 here and one above.

3DSinghVFX replied:
It is added into `setVertexEarStatus` function

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+    else:
+        verticesData[nextIndex].isEar = False
+
+cdef struct Vertex:
+    Py_ssize_t index
+    Vector3 location
+    float angle
+    bint isEar
+    Py_ssize_t previous
+    Py_ssize_t next
```
Move this to the top of the file.

3DSinghVFX replied:
It is moved.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+cdef bint polyCounterClockwise(Vertex* verticesData):
+    cdef Vector3 v1, v2
+    cdef float area = 0.0
+    cdef Py_ssize_t headIndex = 0
```
Head index is not zero. More on this later.

3DSinghVFX replied:
`struct Vertices` is added to track the `head`.
```
cdef struct Vertices:
    Py_ssize_t head
    Vertex *data
```

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+    # Make sure normal (+z) is correct, and the polygon is counter clockwise. Convex vertex has sign >= 0.
+    cdef float sign = (v2.x - v1.x) * (v3.y - v1.y) - (v2.y - v1.y) * (v3.x - v1.x)
+    if sign < 0.0: return -(angle - angle % 0.001)
+    return (angle - angle % 0.001)
```
Why the need for this `angle - angle % 0.001`?

3DSinghVFX replied:
I added this to handle the precision problem. It mainly happens when angles are equal to each other e.g., square polygon.

OmarEmaraDev replied:
Can you elaborate? What happens if the angles are equal?

3DSinghVFX replied:
I have tried without that, now everything is fine.

However, before the angles were changing after some decimal places. May be due to the `pi` that I used before.

3DSinghVFX replied:
It is happening when the input mesh is animated.

3DSinghVFX replied:
Here is an example:
![Peek 2020-06-18 19-46](https://user-images.githubusercontent.com/30294746/85031520-66724200-b19c-11ea-9caf-d25659a011e6.gif)

OmarEmaraDev replied:
I see. Do we need `@cython.cdivision(True)` here since `%` is used?

3DSinghVFX replied:
I think no. `cython -a` shows white color for those lines.
![Screenshot from 2020-06-18 19-58-27](https://user-images.githubusercontent.com/30294746/85033077-27dd8700-b19e-11ea-96e4-f3265cbc2283.png)

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+    cdef float angle
+    cdef Py_ssize_t i, angleMinIndex
+
+    angleMinIndex = 0
+    for i in range(polyLength):
+        if verticesData[i].isEar:
+            angle = abs(verticesData[i].angle)
+            if angleMin > angle:
+                angleMin = angle
+                angleMinIndex = i
```
You are not looping over the linked list correctly. More on this later.

3DSinghVFX replied:
It is changed.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+    cdef Vector3 v2 = verticesData[earIndex].location
+    cdef Vector3 v3 = verticesData[nextIndex].location
+
+    cdef Py_ssize_t i
+    for i in range(polyLength):
+        if i != previousIndex and i != earIndex and i != nextIndex and pointInsideTriangle(v1, v2, v3, verticesData[i].location):
+            return True
+    return False
```
Not looping correctly here either. More on this later.

3DSinghVFX replied:
It is changed.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+# Make sure normal (+z) is correct.
+@cython.cdivision(True)
+cdef bint polyCounterClockwise(Vertex* verticesData):
```
For functions that return bint, name it as `is*`, in this case, `isPolyCounterClockwise`.

3DSinghVFX replied:
Okay. This function is changed to `makeSurePolygonIsCounterClockwise`.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+        verticesData[i].index = oldIndices[polyStart + i]
+        verticesData[i].location = v1
+        verticesData[i].previous = (polyLength + i - 1) % polyLength
+        verticesData[i].next = (i + 1) % polyLength
```
I prefer to handle the first and last vertex independently, to avoid the remainder operator here.

3DSinghVFX replied:
It is done.

3DSinghVFX commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,316 @@
+
+        # Initialization of polygon's vertices.
+        verticesData = <Vertex*>PyMem_Malloc(polyLength * sizeof(Vertex))
+        projectPolygonVertices(vertices, polyStart, polyLength, oldIndices, verticesData)
```
It is done.

3DSinghVFX commented:
> The code is much cleaner now, I think. But you are still having trouble with the linked list, so let me clarify.
> 
> In our case, the doubly liked list is circular, meaning the first element references the last. So the list is fully defined by two values, the array of nodes, which you already have, and a head index, which you don't seem to keep track of. We can put both values in a struct as follows and pass it as a single value to functions:
> 
> ```cython
> cdef struct Vertices:
>     Py_ssize_t head
>     Vertex *data
> ```
> 
> The head index is the index of any of the nodes of the list. When you remove a node, you have to check if it is the head, and if it is, you have to change the head to any other node in the list. In other words, you have to keep track of your head and make sure it is correctly set.
> 
> Secondly, when looping over the linked list, you have two cases: (1) The list is full (No elements were removed) and (2) some elements were removed. If the case is (1), then you may loop over the list just like you would loop over any other list, however, once an element was removed (2), you have to loop over it as a linked list, just like I described before.
> 
> ```python
> currentVertex = headVertex
> while True:
>     # Do something with the current vertex.
>     currentVertex = currentVertex.next
>     if currentVertex == headVertex:
>         break
> ```
> 
> In your case, initialization, projection, and `polyCounterClockwise` all deal with full lists, so you can loop over them normally. In any other case, you have to loop as a linked list. Moreover, a node is removed simply by switching the next and previous indices of its neighbours, don't do anything else, because I see you are setting the ear status, which is unnecessary.

@OmarEmaraDev Thanks for information and clarification.

I have implemented `Vertex` and `Vertices` structures as you suggested above. Based on that I have modified and refined the code.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,353 @@
+            newIndices[polyIndex + 2] = vertexData[vertexData[earIndex].next].index
+            polyIndex += 3
+
+            if earIndex == verticesData.head: verticesData.head = vertexData[earIndex].next
```
Put this line inside the `removeEarVertex` function.

3DSinghVFX replied:
I have tried, it does not work inside the `removeEarVertex` function.

OmarEmaraDev replied:
Can you show me what you tried?

3DSinghVFX replied:
> Put this line inside the `removeEarVertex` function.

Done.

OmarEmaraDev commented on animation_nodes/algorithms/mesh/triangulate_mesh.pyx:
```diff
@@ -0,0 +1,353 @@
+    cdef Vertex* vertexData = verticesData.data
+    cdef Py_ssize_t currentIndex
+
+    cdef float angleMin = 3.14
```
`angleMin` should be the angle of the head vertex.

3DSinghVFX replied:
Okay.

3DSinghVFX replied:
Omar, we can't set it as the angle of the head-vertex because head-vertex may not be an ear but can have small angle.

OmarEmaraDev replied:
I see. Then you should increase the value to 3.15, and add a comment explaining this.

3DSinghVFX replied:
Okay.

3DSinghVFX replied:
Done.
